### DISCUSSION TITLE
Returning shader's source instead of path in custom Material

### DISCUSSION THREAD

**Discussion by tesfabpel** [bevyengine/bevy#5654] [Q&A] [answered]
I'm trying to create a custom Material but I can't figure out how to create a shader given its source code (included at compile-time via the `include_str!` macro) in the `fragment_shader()` function of the `Material` trait. 
That function expects a return type of `ShaderRef` that is an enum including the `Handle` variant but since the function is static (it doesn't take a `self` parameter) and as such, I can't store the Handle as a field as well...

**Comment by tim-blackbird** [comment, marked as answer]
Internally, Bevy and other plugins do something like this.
```rs
const SHADER_HANDLE: HandleUntyped =
	HandleUntyped::weak_from_u64(Shader::TYPE_UUID, /* insert a large random number here */);

// You can use this helper macro as suggested by komadori below :)
load_internal_asset!(
    app,
    SHADER_HANDLE,
    "shader/path.wgsl",
    Shader::from_wgsl
);

// Which expands to the following code.
let mut shaders = app.world.resource_mut::<Assets<Shader>>();
shaders.set_untracked(SHADER_HANDLE, Shader::from_wgsl(include_str!("shader/path.wgsl")));
```
And then return this handle from `fragment_shader()`.

**Reply by komadori** [reply]
@devil-ira is correct, but better still Bevy exports the load_internal_asset macro to encapsulate the boilerplate:

https://docs.rs/bevy/latest/bevy/asset/macro.load_internal_asset.html

You still have to think of a large 64-bit random number yourself though 😋.

**Reply by tesfabpel** [reply]
Thank you! I'll try this later!

**Reply by tesfabpel** [reply]
Yes, it works ok! Thank you all again! 
I find this way a little bit dirty... You have to use a "random" large random number (instead of having bevy assign one for you). Hopefully in a future version there will be a way to register those assets differently...
For example, a game may want to support loading shaders from Steam Workshop (like I believe GMod does) and this static approach is a bit constraining probably, since I can't create a custom material with parameters (the `fragment_shader` and the `vertex_shader` functions are parameter-less static) forbidding me to load arbitrary shader files from disk / memory / whatever...
